src/stream/src/executor/expand.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::Debug;
+
+use futures::StreamExt;
+use futures_async_stream::try_stream;
+use risingwave_common::array::PrimitiveArray;
+use risingwave_common::catalog::{Field, Schema};
+use risingwave_common::types::DataType;
+
+use super::error::StreamExecutorError;
+use super::*;
+
+pub struct ExpandExecutor {
+    input: BoxedExecutor,
+    schema: Schema,
+    pk_indices: PkIndices,
+    expanded_keys: Vec<Vec<usize>>,
+}
+
+impl ExpandExecutor {
+    pub fn new(
+        input: Box<dyn Executor>,
+        pk_indices: PkIndices,
+        expanded_keys: Vec<Vec<usize>>,
+    ) -> Self {
+        let mut schema = input.schema().to_owned();
+        schema
+            .fields
+            .push(Field::with_name(DataType::Int64, "flag"));
+        Self {
+            input,
+            schema,
+            pk_indices,
+            expanded_keys,
+        }
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    async fn execute_inner(self) {
+        #[for_await]
+        for msg in self.input.execute() {
+            let msg = msg?;
+            match msg {
+                Message::Chunk(chunk) => {
```

Comment:
nit: avoid deep indentation with early return:
```
let chunk = match msg {
  Message::Chunk(chunk) => chunk,
  m => {yield m; continue}
};
```

Note:
* `let_else` is syntactically the same but seems problematic: #3270